File: Spark_FastQ_Scala/src/main/scala/PerBaseQual/FastQ_PerBaseQual_Reducer.scala
package main.scala.PerBaseQual

import main.scala.utils.AvgCount

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 02.01.16.
  */
object FastQ_PerBaseQual_Reducer {

  def createAverageCount(value: Double): AvgCount = {
    new AvgCount(value,1)
  }

  def addAndCount(a: AvgCount, x: Double): AvgCount = {
    a.incrementTotalBy(x)
    a.incrementNumBy(1)
    a
  }

  def combine(a: AvgCount, b: AvgCount): AvgCount = {
    a.incrementTotalBy(b.getTotal())
    a.incrementNumBy(b.getNum())
    a
  }

}

File: MapReduce_FastQ_BaseQual_Baseline/src/main/java/baseQual/FastQ_PerBaseQual_Reducer.java
package baseQual;

import org.apache.hadoop.io.DoubleWritable;
import org.apache.hadoop.io.IntWritable;
import utils.IdentifierPositionKeyWritable;

import java.io.IOException;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 21.04.2016
 */
public class FastQ_PerBaseQual_Reducer extends org.apache.hadoop.mapreduce.Reducer<IdentifierPositionKeyWritable, IntWritable, IdentifierPositionKeyWritable, DoubleWritable> {

    @Override
    protected void reduce(IdentifierPositionKeyWritable key, Iterable<IntWritable> values, Context context) throws IOException, InterruptedException {
        context.write(key, getMeanValue(values));
    }

    private DoubleWritable getMeanValue(Iterable<IntWritable> values) {
        int numberOfQualityValues = 0;
        double sumOfQualityValues = 0;
        for (IntWritable qualVal : values) {
            sumOfQualityValues += qualVal.get();
            numberOfQualityValues++;
        }
        return new DoubleWritable(sumOfQualityValues / numberOfQualityValues);
    }

}